### PR TITLE
PHPLIB-518: Provide MongoDB\Collection::rename() method to rename a collection

### PR DESCRIPTION
**Pull request by tanlisu**
https://jira.mongodb.org/browse/PHPLIB-518

### CODE REVIEW

File: docs/reference/method/MongoDBCollection-renameCollection.txt
Comment (tanlisu) [inline docs/reference/method/MongoDBCollection-renameCollection.txt:74]: copied from `MongoDBCollection-dropCollection.txt`, I'm not sure what the correct output should be yet

File: tests/Operation/RenameCollectionFunctionalTest.php
Comment (tanlisu) [inline tests/Operation/RenameCollectionFunctionalTest.php:22]: I was originally using `$this->getDatabaseName()` for the `$databaseName` parameter, but it gave the error `MongoDB\Driver\Exception\CommandException: renameCollection may only be run against the admin database` so I changed all the tests to use the `'admin'` database - not sure if that was the right thing to do.

File: tests/Operation/RenameCollectionFunctionalTest.php
Comment (tanlisu) [inline tests/Operation/RenameCollectionFunctionalTest.php:49]: The tests aren’t passing now because the assertion in this line fails; I printed the `$commandResult` object for the error message but am not sure what’s wrong still.

Screen Shot 2021-07-07 at 4 46 25 PM

File: tests/Operation/RenameCollectionFunctionalTest.php
Comment (jmikola) [reply tests/Operation/RenameCollectionFunctionalTest.php:22]: The [server manual]((https://docs.mongodb.com/manual/reference/command/renameCollection/) does state that the command itself must be run against the `admin` database:

> Issue the renameCollection command against the admin database.

However, that shouldn't limit the source and target namespaces involved in the renaming.

File: src/Operation/RenameCollection.php
Comment (jmikola) [inline src/Operation/RenameCollection.php:150]: This where you should use `admin`.

File: docs/reference/method/MongoDBCollection-rename.txt
Comment (tanlisu) [reply docs/reference/method/MongoDBCollection-rename.txt:70]: Yup, the only fields in the `rename` response besides "ok" were "$clusterTime" and "operationTime", but I saw that those were omitted in `MongoDBCollection-drop.txt` so I left them out here as well.

Comment (jmikola) [reply docs/reference/method/MongoDBCollection-rename.txt:70]: Indeed, those are generic fields that the server adds to everything (re: [causal consistency](https://github.com/mongodb/specifications/blob/master/source/causal-consistency/causal-consistency.rst)), so it's reasonable to omit them from our examples.

File: tests/Operation/RenameCollectionFunctionalTest.php
Comment (jmikola) [inline tests/Operation/RenameCollectionFunctionalTest.php:124]: This was presumably copied from `MongoDB\Tests\Operation\FunctionalTestCase::tearDown()`. I think this behavior is actually incorrect as it means we'll never call the `tearDown()` method of the top-level MongoDB\Tests\FunctionalTestCase` class, which disables fail points.

I opened [PHPLIB-693](https://jira.mongodb.org/browse/PHPLIB-693) to track that, so this can remain as-is for now.

File: tests/Operation/RenameCollectionFunctionalTest.php
Comment (jmikola) [inline tests/Operation/RenameCollectionFunctionalTest.php:14]: I think this file would do well to have a test for the `dropTarget` option. Perhaps two cases:

* Rename an existing collection when the target already exists and expect a server error of some kind.
* Repeat the first test but specify `dropTarget` as true and expect success.

File: docs/reference/method/MongoDBDatabase-renameCollection.txt
Comment (tanlisu) [reply docs/reference/method/MongoDBDatabase-renameCollection.txt:22]: I don't think it works to determine whether a collection name or namespace is being provided by checking whether the argument contains a dot, since the collection names used in the tests contain a dot as well. So I'm not sure how it would be possible to do this actually - any ideas?

File: tests/Collection/CollectionFunctionalTest.php
Comment (tanlisu) [reply tests/Collection/CollectionFunctionalTest.php:17]: I was using `DropCollection` and `FindOne` for the renamed collection in `testRename()`, so I don't think I'm able to use the Collection class methods in this case right?

Comment (jmikola) [reply tests/Collection/CollectionFunctionalTest.php:17]: In a later comment, I suggest creating a Collection instance for the renamed class. That would be taboo in the Operation test classes since those shouldn't concern themselves with the Collection class, but it shouldn't be a concern in CollectionFunctionalTest since we're already calling Collection methods elsewhere in this class.

File: docs/reference/method/MongoDBDatabase-renameCollection.txt
Comment (jmikola) [reply docs/reference/method/MongoDBDatabase-renameCollection.txt:22]: Very good point. This would work the other way around (e.g. `$databaseNameOrNamespace`) because database names cannot contain dots; however, it's ambiguous if we're trying to differentiate a namespace from a collection name.

In that case, I think it makes sense to leave `$toNamespace` as-is.

`$fromNamespace` can still be changed to `$collectionName` for consistency with `Database::dropCollection()`, and the helper method can build the namespace to send to the Operation internally.

Comment (jmikola) [reply docs/reference/method/MongoDBDatabase-renameCollection.txt:22]: To add on to this, let's change the API of both the helpers _and_ Operation class to no longer take namespace strings as paramaters:

```
$database->renameCollection($collectionName, $toDatabaseName, $toCollectionName);
$collection->rename($toDatabaseName, $toCollectionName);
```

The Operation class will then have four required parameters:

 * fromDatabaseName
 * fromCollectionName
 * toDatabaseName
 * toCollectionName

It can then build the namespace internally with string concatenation in the constructor (allowing you to keep the internal private properties that store namespaces).

For the Database helper, fromDatabaseName would be pulled from the database instance's own name. And for the Collection helper, both "from" parameters would be pulled from the collection instance.

File: tests/Database/DatabaseFunctionalTest.php
Comment (tanlisu) [reply tests/Database/DatabaseFunctionalTest.php:232]: Oh I used `$this->assertCollectionCount()` because I saw that the `testDrop()` function used it, but I realize now that function is testing dropping a database, not a collection. I'll change `testRenameCollection()` to use the existence assertions.

File: src/Operation/RenameCollection.php
Comment (alcaeus) [reply src/Operation/RenameCollection.php:126]: Defensive coding practices would suggest leaving this in. Creating a drivers ticket to check for this makes sense to me.

File: tests/FunctionalTestCase.php
Comment (jmikola) [inline tests/FunctionalTestCase.php:193]: Pending @alcaeus' input, this can be changed to either `?string` or `string|null`.

File: tests/Operation/RenameCollectionFunctionalTest.php
Comment (jmikola) [inline tests/Operation/RenameCollectionFunctionalTest.php:121]: Why is [`@depends`](https://phpunit.readthedocs.io/en/9.5/annotations.html#depends) used here? This doesn't seem like it would depend on `testRenameExistingCollection`.

File: tests/Operation/RenameCollectionFunctionalTest.php
Comment (jmikola) [inline tests/Operation/RenameCollectionFunctionalTest.php:125]: I think this is redundant and can be removed. The source collection is dropped by the parent `setUp()` method in `MongoDB\Tests\Operation\FunctionalTestCase`, which will run before each test.

If the collection actually existed and wasn't dropped, then we'd expect this test to fail anyway because the server would no longer throw a NamespaceNotFound(26) error.

File: tests/Operation/RenameCollectionFunctionalTest.php
Comment (jmikola) [inline tests/Operation/RenameCollectionFunctionalTest.php:134]: This looks like it was copied from `testDefaultWriteConcernIsOmitted()`, but it's not relevant to this test. I suggest removing it, since it won't be sent in the command anyway.

File: tests/Operation/RenameCollectionFunctionalTest.php
Comment (jmikola) [inline tests/Operation/RenameCollectionFunctionalTest.php:94]:
```suggestion
        $this->assertSameDocument(['_id' => 1], $operation->execute($server));
```

Remember: `FindOne::execute()` returns a document, not a cursor.

File: tests/Operation/RenameCollectionFunctionalTest.php
Comment (jmikola) [inline tests/Operation/RenameCollectionFunctionalTest.php:72]:
```suggestion
    public function testRenameCollectionToNonexistentTarget(): void
```

A nonexistent source collection is an exception, and covered in a later test that expects an error. The successful path involves a collection that _does_ exist, so I don't think we need to clarify that. The distinguishing characteristic of this test is that the target does not exist.

File: tests/Operation/RenameCollectionFunctionalTest.php
Comment (jmikola) [inline tests/Operation/RenameCollectionFunctionalTest.php:97]:
```suggestion
    public function testRenameCollectionExistingTarget(): void
```

File: tests/Operation/RenameCollectionFunctionalTest.php
Comment (jmikola) [inline tests/Operation/RenameCollectionFunctionalTest.php:117]:
```suggestion
        $operation->execute($server);
```

Even if the operation doesn't throw for some reason (and the test ultimately fails), this return value is never used. Consider what was done in the `testDefaultWriteConcernIsOmitted` and session tests.

File: tests/Operation/RenameCollectionFunctionalTest.php
Comment (jmikola) [inline tests/Operation/RenameCollectionFunctionalTest.php:136]:
```suggestion
        $operation->execute($this->getPrimaryServer());
```

**Review by jmikola — COMMENTED** [review]
Reviewed the code files but skipped the docs on this round. I'll come back to those on Friday and do a local build to check for any RST errors.

File: tests/Collection/CollectionFunctionalTest.php
Comment (tanlisu) [reply tests/Collection/CollectionFunctionalTest.php]: I tried creating the target database explicitly as you suggested; the PHPUnit tests with sharded cluster seem to be working, but the evergreen sharded cluster test doesn't. Do you know what might be wrong or should I just mark the the test as skipped for now?

File: src/Operation/RenameCollection.php
Comment (alcaeus) [reply src/Operation/RenameCollection.php:81]: Yes, any new API should be as strictly typed as possible with the current PHP version constraints (scalar types, nullable types, `void` allowed, `object`, property types, union types, `mixed` not allowed).

File: tests/FunctionalTestCase.php
Comment (alcaeus) [reply tests/FunctionalTestCase.php]: I believe PHPDoc understands `?string`, but since that syntax was only introduced with PHP 7.1, many people still prefer `string|null`. I have no preference for either since I prefer using an argument type directly and skipping the `@param` annotation since the type information is already in the signature.

File: src/Collection.php
Comment (alcaeus) [reply src/Collection.php:1013]: In this case the parameter should be documented as nullable. As for the rest of the documentation, I'll defer to your judgement, it's better than mine in 99.99999% of all cases.

File: src/Operation/RenameCollection.php
Comment (jmikola) [reply src/Operation/RenameCollection.php:81]: @tanlisu please apply these changes.

File: src/Collection.php
Comment (jmikola) [reply src/Collection.php:1013]: @tanlisu Please change type to `?string` and apply the following suggested change to the description.

File: src/Database.php
Comment (jmikola) [reply src/Database.php:480]: @tanlisu Please change type to ?string and apply the following suggested change to the description.

File: tests/FunctionalTestCase.php
Comment (jmikola) [reply tests/FunctionalTestCase.php]: @tanlisu: Let's just delete both of these `@param` lines.

File: tests/Collection/CollectionFunctionalTest.php
Comment (jmikola) [reply tests/Collection/CollectionFunctionalTest.php]: > PHPUnit tests with sharded cluster seem to be working, but the evergreen sharded cluster test doesn't

What do you mean by "PHPUnit tests with sharded cluster"? Were you running a sharded cluster locally?

File: docs/includes/apiargs-MongoDBCollection-method-rename-param.yaml
Comment (jmikola) [inline docs/includes/apiargs-MongoDBCollection-method-rename-param.yaml:8]: This text is leftover from when we had a single parameter for the collection name or namespace. I suggest replacing it with:

> The new name of the collection.

**Review by jmikola — COMMENTED** [review]
Built the documentation locally and didn't see any syntax errors. Left one suggestion to reword, and then some other suggestions for outstanding comments on code docs.

File: docs/reference/method/MongoDBCollection-rename.txt
Comment (jmikola) [inline docs/reference/method/MongoDBCollection-rename.txt:5]:
```suggestion
.. versionadded:: 1.10
```

File: docs/reference/method/MongoDBDatabase-renameCollection.txt
Comment (jmikola) [inline docs/reference/method/MongoDBDatabase-renameCollection.txt:5]:
```suggestion
.. versionadded:: 1.10
```

File: tests/Collection/CollectionFunctionalTest.php
Comment (tanlisu) [reply tests/Collection/CollectionFunctionalTest.php]: > Are you referring to the GitHub CI job that uses a sharded cluster (https://github.com/mongodb/mongo-php-library/runs/3125307713)?

Yes, that's what I meant. Got it!

File: tests/FunctionalTestCase.php
Comment (jmikola) [inline tests/FunctionalTestCase.php:189]: Please remove the `@param` line for `$callback` as well. It adds nothing of value.

File: tests/FunctionalTestCase.php
Comment (jmikola) [inline tests/FunctionalTestCase.php:158]:
```suggestion
     * $databaseName defaults to TestCase::getDatabaseName() if unspecified.
```

File: tests/FunctionalTestCase.php
Comment (jmikola) [inline tests/FunctionalTestCase.php:183]:
```suggestion
     *
     * $databaseName defaults to TestCase::getDatabaseName() if unspecified.
```

**Review by jmikola — APPROVED** [review]
Tests appear to be passing, which confirms that the earlier issue was unrelated.

LGTM after a few doc block suggestions.

**Comment by jmikola** [comment]
When squashing and merging this, I'd suggest the following description.

> PHPLIB-518: RenameCollection operation
>
> Adds Database::renameCollection() and Collection::rename() methods.
>
> Move assertCollectionDoesNotExist() and assertCollectionExists() to FunctionalTestCase

Most of the other commit messages shouldn't be relevant, and this summarizes the actual changes we might want to look back on down the line. You can also rename the PR itself to "PHPLIB-518: RenameCollection operation", which should suggest that title when squashing.

**Pull request merged**
